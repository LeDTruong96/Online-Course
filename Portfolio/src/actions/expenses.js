import { v4 as uuidv4 } from 'uuid';

// ADD_EXPENSE
export const addExpenses = ({ description = '', note = '', amount = 0, createAt = 0 } = {}) => {
	return {
		type: 'ADD_EXPENSE',
		expenses: {
			id: uuidv4(),
			description,
			note,
			amount,
			createAt
		}
	};
};

// REMOVE_EXPENSE
export const removeExpenses = ({ id } = {}) => {
	return {
		type: 'REMOVE_EXPENSE',
		id
	};
};

// EDIT_EXPENSE
export const editExpenses = (id, expense = {}) => {
	return {
		type: 'EDIT_EXPENSE',
		id,
		expense
	};
};
