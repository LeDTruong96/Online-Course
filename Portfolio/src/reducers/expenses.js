// Expense reducers
const expenseReducerDefaultValue = [];

export default (state = expenseReducerDefaultValue, action) => {
	switch (action.type) {
		case 'ADD_EXPENSE':
			return [ ...state, action.expenses ];
		case 'REMOVE_EXPENSE':
			return state.filter(({ id }) => id !== action.id);
		case 'EDIT_EXPENSE':
			return state.map((expense) => {
				if (expense.id === action.id) {
					return { ...expense, ...action.expense };
				} else {
					return expense;
				}
			});
		default:
			return state;
	}
};
