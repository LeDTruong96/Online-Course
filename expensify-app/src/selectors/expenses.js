// Get visible expenses
export default ({ expenses, filters }) => {
	return expenses
		.filter((expense) => {
			const matchStartDate =
				typeof filters.startDate !== 'number' || expense.createAt >= filters.startDate;
			const matchEndDate =
				typeof filters.endDate !== 'number' || expense.createAt <= filters.endDate;
			const matchText = expense.description
				.toLowerCase()
				.includes(filters.text.toLowerCase());

			return matchStartDate && matchEndDate && matchText;
		})
		.sort((fExpense, sExpense) => {
			if (filters.sortBy === 'date') {
				return fExpense.createAt < sExpense.createAt ? 1 : -1;
			} else if (filters.sortBy === 'amount') {
				return fExpense.amount < sExpense.amount ? 1 : -1;
			}
		});
};
