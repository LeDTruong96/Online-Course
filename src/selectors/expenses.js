// Get visible expenses
import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
	return expenses
		.filter((expense) => {
			const createAt = moment(expense.createdAt);
			const startDateMatch = !!startDate ? createAt.isSameOrAfter(startDate) : true;
			const endDateMatch = !!endDate ? createAt.isSameOrBefore(endDate) : true;
			const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

			return startDateMatch && endDateMatch && textMatch;
		})
		.sort((a, b) => {
			if (sortBy === 'date') {
				return a.createdAt < b.createdAt ? 1 : -1;
			} else if (sortBy === 'amount') {
				return a.amount < b.amount ? 1 : -1;
			}
		});
};
