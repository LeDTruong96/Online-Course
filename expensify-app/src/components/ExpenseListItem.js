import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';
import { Link } from 'react-router-dom';
import moment from 'moment';

const ExpenseListItem = ({ id, description, amount, createdAt }) => {
	return (
		<div>
			<Link to={`/edit/${id}`}>
				<h3>{description}</h3>
			</Link>
			<p>
				{amount} - {moment(createdAt).format('MMM Do, YYYY')}
			</p>
		</div>
	);
};

export default ExpenseListItem;
