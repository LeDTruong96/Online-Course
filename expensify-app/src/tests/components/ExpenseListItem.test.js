import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('ExpenseList', () => {
	const expense = {
		...expenses[1],
		id: 'id_element_of_expenses'
	};
	const wrapper = shallow(<ExpenseListItem {...expense} />);
	expect(wrapper).toMatchSnapshot();
});
