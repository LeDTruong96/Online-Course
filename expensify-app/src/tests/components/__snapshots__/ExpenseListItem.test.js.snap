// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList 1`] = `
<div>
  <Link
    to="/edit/id_element_of_expenses"
  >
    <h3>
      Rent
    </h3>
  </Link>
  <p>
    109500
     - 
    Dec 28th, 1969
  </p>
</div>
`;
